tests/AstrolabeWorkloadExecutor/AstrolabeTestRunner.cs

```diff
@@ -0,0 +1,174 @@
+using MongoDB.Bson.TestHelpers.JsonDrivenTests;
+using MongoDB.Driver;
+using MongoDB.Driver.Core;
+using MongoDB.Driver.TestHelpers;
+using MongoDB.Driver.Tests.JsonDrivenTests;
+using MongoDB.Driver.Tests.Specifications.Runner;
+
+namespace WorkloadExecutor
+{
+    public class AstrolabeTestRunner : MongoClientJsonDrivenTestRunnerBase
+    {
+        // private fields
+        private readonly CancellationToken _cancellationToken;
+        private readonly Action _incrementOperationSuccesses;
+        private readonly Action _incrementOperationErrors;
+        private readonly Action _incrementOperationFailures;
+
+        // protected properties
+        protected override string[] ExpectedSharedColumns => new[] { "_path", "database", "collection", "testData", "tests" };
+        protected override string[] ExpectedTestColumns => new[] { "operations", "async" };
+
+        public AstrolabeTestRunner(
+            Action incrementOperationSuccesses,
+            Action incrementOperationErrors,
+            Action incrementOperationFailures,
+            CancellationToken cancellationToken)
+        {
+            _incrementOperationSuccesses = incrementOperationSuccesses;
+            _incrementOperationErrors = incrementOperationErrors;
+            _incrementOperationFailures = incrementOperationFailures;
+            _cancellationToken = cancellationToken;
+        }
+
+        protected override string DatabaseNameKey => "database";
+        protected override string CollectionNameKey => "collection";
+        protected override string DataKey => "testData";
+
+        // public methods
+        public void Run(JsonDrivenTestCase testCase)
+        {
+            SetupAndRunTest(testCase);
+        }
+
+        // protected methods
+        protected override void ExecuteOperations(IMongoClient client, Dictionary<string, object> objectMap, BsonDocument test, EventCapturer eventCapturer = null)
+        {
+            _objectMap = objectMap;
+
+            var factory = new JsonDrivenTestFactory(client, DatabaseName, CollectionName, bucketName: null, objectMap, eventCapturer);
+
+            Func<JsonDrivenTest, AstrolabeJsonDrivenTest> wrapTest = wrapped =>
+                new AstrolabeJsonDrivenTest(wrapped, _incrementOperationSuccesses, _incrementOperationErrors, _incrementOperationFailures);
+
+            foreach (var operation in test[OperationsKey].AsBsonArray.Cast<BsonDocument>())
+            {
+                if (_cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                var receiver = operation["object"].AsString;
+                var name = operation["name"].AsString;
+                JsonDrivenTest jsonDrivenTest;
+                try
+                {
+                    var innerTest = factory.CreateTest(receiver, name);
+                    jsonDrivenTest = wrapTest(innerTest);
+                }
+                catch (FormatException)
```

Comment:
this looks like something which requires a different solution...

Reply:
Done.